torch/csrc/autograd/variable.h

```diff
@@ -216,16 +217,25 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
   // must be protected by mutex_
   std::shared_ptr<ForwardGrad> fw_grad_;
 
-  std::vector<std::shared_ptr<FunctionPreHook>> hooks_;
+  // The hooks_ field is actually reused by both python and cpp logic
+  // For both cases, we have a data structure, cpp_hooks_list_ (cpp)
+  // or dict (python) which is the canonical copy.
+  // Then, for both cases, we always register a single hook to
+  // hooks_ which wraps all the hooks in the list/dict.
+  // And, again in both cases, if the grad_fn exists on that tensor
+  // we will additionally register a single hook to the grad_fn.
+  //
+  // Note that the cpp and python use cases aren't actually aware of
+  // each other, so using both is not defined behavior.
```

Comment:
What would happen in practice before this PR? They will just run in arbitrary order?
What about after this PR?

Reply:
From the python side the hooks_ field is cleared everytime backward_hooks field is set, and a PyFunctionHook is registered.
From the cpp side, hooks_ field is set whenever a cpp hook is registered AND cpp_hooks_list_ is empty. (In the case where we register a retains_grad hook that is also the case, but it shouldn't matter because hooks_ is only used on leaf nodes and retains_grad is no-op when invoked on leaf node)

What happens is that one will be overwritten by the other.

This is still the case after this PR, but I'm okay with not caring about mixed Python + cpp case.

Reply:
Ho wow this is very badly broken :O good to know!